### PR TITLE
Phase 6: Per-Foil Physics Normalization — Fix Aft-Foil Cp Denominator

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current `_umag_q` function computes a **single** domain-mean Umag from ALL mesh nodes (fore + aft + wake). For tandem samples, the aft foil sits deep in the fore-foil's wake where local velocities are significantly lower than freestream. This means the `q = 0.5 * Umag²` denominator used for aft-foil Cp is **physically incorrect** — it uses the global (inflated) dynamic pressure instead of the local aft-foil reference.

The consequence: aft-foil Cp values in the training targets are systematically inflated relative to what local physics implies. The model must implicitly learn to undo this normalization artifact, wasting representational capacity. Single-foil samples are completely unaffected (no wake interference).

**Fix:** For tandem samples, split the normalization so:
- **Fore-foil nodes** use `q_fore` (computed from fore-foil nodes only)
- **Aft-foil nodes** use `q_aft` (computed from aft-foil nodes only)

This is a physically motivated bias correction applied consistently to both training and validation. No new hyperparameters.

**Expected impact:** -2% to -5% p_tan. The model has been fighting an incorrect normalization for aft-foil nodes from day one.

**Risk:** MEDIUM-LOW. If `Umag_aft ≈ Umag_fore` (wide gap, weak wake), the split degenerates cleanly to current behavior. Edge case: very tight gap → very low `q_aft` → need a minimum clamp (see Step 4).

### Step 1: Add config flag

In `Config` dataclass (~line 950):
```python
per_foil_pnorm: bool = False
```

In argparse:
```python
parser.add_argument('--per_foil_pnorm', action='store_true',
                    help='Per-foil Cp normalization: separate q for fore/aft nodes in tandem')
```

### Step 2: Add helper function (after `_umag_q`, ~line 1030)

```python
def _umag_q_perfoil(y, fore_mask, aft_mask):
    """Separate q normalization for fore and aft foil nodes in tandem samples."""
    Umag_fore, q_fore = _umag_q(y, fore_mask)
    Umag_aft, q_aft = _umag_q(y, aft_mask)
    # Clamp aft q to avoid blow-up in deep-wake scenarios
    q_aft = q_aft.clamp(min=0.1 * q_fore)
    return Umag_fore, q_fore, Umag_aft, q_aft
```

### Step 3: Locate `_aft_foil_mask` in the training loop

Search for `_aft_foil_mask` in `train.py` — it should be computed around line 1562-1566 from raw (pre-standardization) x. Confirm this mask correctly identifies aft-foil nodes for tandem samples.

Also find the `is_tandem` indicator — look for `x[:, 0, 21]` (gap feature, nonzero for tandem) or `_is_tandem_batch`.

### Step 4: Apply per-foil normalization in training and validation loops

Find where `Umag, q = _umag_q(y, mask)` is called in the training loop, then add AFTER it:

```python
if cfg.per_foil_pnorm and _aft_foil_mask is not None:
    _fore_mask = mask & ~_aft_foil_mask           # [B, N] fore-foil + non-tandem nodes
    _aft_mask_valid = mask & _aft_foil_mask       # [B, N] aft-foil nodes
    _is_tandem_b = (x[:, 0, 21].abs() > 0.01)    # [B] tandem indicator (pre-standardization)
    
    # For each tandem sample, override q with per-foil values
    # Build per-sample q tensors [B, 1, 1] for use in _phys_norm
    q_effective = q.clone()   # start with global q
    Umag_effective = Umag.clone()
    
    for b in range(y.shape[0]):
        if _is_tandem_b[b] and _fore_mask[b].any() and _aft_mask_valid[b].any():
            _, q_f, _, q_a = _umag_q_perfoil(
                y[b:b+1], _fore_mask[b:b+1], _aft_mask_valid[b:b+1]
            )
            # Apply split: fore nodes use q_fore, aft nodes use q_aft
            # You'll need to adapt _phys_norm to accept per-node q, OR
            # apply it block by block:
            # e.g. y_norm[b, _fore_mask[b]] = _phys_norm(y[b:b+1, _fore_mask[b]], ..., q_f)
            # e.g. y_norm[b, _aft_mask_valid[b]] = _phys_norm(y[b:b+1, _aft_mask_valid[b]], ..., q_a)
```

**Study `_phys_norm` and `_umag_q` carefully before implementing** — the exact API depends on their signature. The key invariant: fore-foil pressure nodes normalized with `q_fore`, aft-foil pressure nodes with `q_aft`.

**CRITICAL:** Apply the same split in the **validation loop** (find the second `_umag_q` / `_phys_norm` call, ~line 2090-2095). Train/val normalization must match.

### Step 5: Sanity check at epoch 0

```python
if cfg.per_foil_pnorm and epoch == 0 and step == 0:
    for b in range(min(4, y.shape[0])):
        if _is_tandem_b[b] and _fore_mask[b].any() and _aft_mask_valid[b].any():
            _, q_f, _, q_a = _umag_q_perfoil(y[b:b+1], _fore_mask[b:b+1], _aft_mask_valid[b:b+1])
            print(f"[per_foil_pnorm] b={b}: q_fore={q_f.item():.4f}, q_aft={q_a.item():.4f}, ratio={q_a.item()/q_f.item():.3f}")
```

Expected: `q_aft / q_fore` typically 0.4-0.8 for tandem samples. If ratio ≈ 1.0, the mask may be wrong.

### Step 6: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/per-foil-pnorm-s42" --wandb_group phase6/per-foil-pnorm \
  --per_foil_pnorm --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aft_foil_srf_context

# Seed 73
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/per-foil-pnorm-s73" --wandb_group phase6/per-foil-pnorm \
  --per_foil_pnorm --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aft_foil_srf_context
```

Run both in parallel. Report W&B run IDs. 

## Baseline

Current single-model baseline (PR #2127 — AftSRF KNN Context K=8):

| Metric | 2-seed avg | Target to beat |
|--------|------------|----------------|
| p_in | **13.02** | < 13.02 |
| p_oodc | **7.62** | < 7.62 |
| **p_tan** | **29.91** | **< 29.91** |
| p_re | **6.47** | < 6.47 |

W&B: zosxwjmm (seed 42, p_tan=29.96), twilqf1x (seed 73, p_tan=29.87)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-aft-srf-ctx" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aft_foil_srf_context
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results: Per-Foil Physics Normalization

### Configuration
- `--per_foil_pnorm` with all baseline flags (`--aft_foil_srf --aug_gap_stagger_sigma 0.02 --aft_foil_srf_context`)
- Seeds 42 and 73, 133 epochs reached (180-min timeout)
- W&B group: `phase6/per-foil-pnorm`

### Results

| Metric | Baseline (2-seed) | s42 | s73 | 2-seed avg | Δ |
|--------|-------------------|-----|-----|-----------|---|
| p_in   | **13.02**         | 14.8 | 16.2 | 15.50 | **+19.0%** |
| p_tan  | **29.91**         | 30.9 | 31.7 | 31.30 | **+4.6%** |
| p_oodc | **7.62**          | 8.3 | 8.6 | 8.45 | **+10.9%** |
| p_re   | **6.47**          | 7.4 | 6.9 | 7.15 | **+10.5%** |

### W&B Runs
- Seed 42: `bc8yxmb6` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/bc8yxmb6
- Seed 73: `urq07o3e` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/urq07o3e

### Analysis

**Clear negative result — all metrics regressed significantly.**

**Root causes:**

1. **Distribution shift from phys_stats mismatch**: The `phys_stats` (mean/std for z-score normalization) were computed using global q. When per-foil q is applied, aft-foil Cp values are rescaled (divided by smaller q_aft), creating larger Cp magnitudes. These larger values are then z-scored using stats computed for the original (smaller) Cp range, leading to a systematic bias in the target distribution for ~30% of training samples.

2. **Training throughput loss**: The per-sample for-loop in `_build_perfoil_q` slowed training — only reached epoch 133 vs 155 for baseline (14% fewer training steps). This compounds the already-worse loss landscape.

3. **p_in regression is worst (+19%)**: Single-foil samples should be completely unaffected by per-foil q (no aft-foil nodes). The large p_in regression suggests the phys_stats mismatch during training destabilized learning globally, not just for tandem samples.

**Why the hypothesis was wrong:**
The original hypothesis assumed that aft-foil nodes are "incorrectly" normalized with global q. However, the model has already learned to compensate for this in the current setup — the physics normalization is a fixed linear transform, and the model's learned weights implicitly account for it. Changing the normalization mid-approach (without recomputing phys_stats) creates a mismatch that hurts rather than helps.

**What would need to change for this to work:**
- Recompute `phys_stats` with per-foil q applied during the stats computation pass
- Vectorize `_build_perfoil_q` to eliminate the per-sample loop (use masked scatter/gather ops)
- Consider applying per-foil q only to the loss computation (not to the model's input targets), so the model sees consistent targets but is evaluated with physically correct metrics

### Suggested Follow-ups
- **Per-foil q in loss only**: Keep global q for target normalization, but use per-foil q when computing validation metrics. This tests whether the metric improvement is real without disrupting training.
- **Recomputed phys_stats**: If the direction is worth pursuing, recompute phys_stats with per-foil q to eliminate the distribution shift.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2136 because ADVISOR: Closing — all metrics regressed. Excellent diagnostic work, thorfinn. Your root cause analysis is spot-on.

## Results Summary

| Metric | Baseline | per_foil_pnorm 2-seed avg | Delta |
|--------|----------|--------------------------|-------|
| p_in | 13.02 | 15.50 | +19.0% ❌ |
| p_oodc | 7.62 | 8.45 | +10.9% ❌ |
| p_tan | 29.91 | 31.30 | +4.6% ❌ |
| p_re | 6.47 | 7.15 | +10.5% ❌ |

## Why It Failed

Your diagnosis is correct on all counts:

1. **phys_stats mismatch:** The normalization stats were computed with global q. Switching to per-foil q at train time shifts the Cp distribution for ~30% of samples without updating the z-score stats → systematic bias.

2. **Training slowdown:** The per-sample for-loop cost you 22 epochs (133 vs 155). This is a recurring issue — any per-sample computation is expensive.

3. **p_in regression confirms global damage:** Single-foil should be unaffected, yet p_in +19% → the normalization mismatch destabilized learning globally.

## Lessons Learned

The model has already learned to compensate for the 'incorrect' global q through its weights. The normalization is a fixed linear transform — the model handles it fine implicitly. Changing the normalization without recomputing phys_stats introduces an asymmetry between training distribution and what the model expects, which is worse than the original 'incorrect' normalization.

This is a generalizable finding: **don't change data normalization without recomputing all derived statistics.** The idea of per-foil q in loss-only (your suggestion) is sound but I want to redirect your time to a higher-priority direction.

New assignment coming..
